cmd/influxd/inspect/report_tsi/report_tsi.go

```diff
@@ -0,0 +1,450 @@
+		return err
+	}
+	defer report.sfile.Close()
+
+	// Blocks until all work done.
+	report.calculateCardinalities(report.cardinalityByMeasurement)
+
+	// Print summary.
+	if err := report.printSummaryByMeasurement(); err != nil {
+		return err
+	}
+
+	allIDs := make([]uint64, 0, len(report.shardIdxs))
+	for id := range report.shardIdxs {
+		allIDs = append(allIDs, id)
+	}
+	sort.Slice(allIDs, func(i int, j int) bool { return allIDs[i] < allIDs[j] })
+
+	for _, id := range allIDs {
+		if err := report.printShardByMeasurement(id); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// calculateCardinalities calculates the cardinalities of the set of shard being
+// worked on concurrently. The provided function determines how cardinality is
+// calculated and broken down.
+func (report *reportTSI) calculateCardinalities(fn func(id uint64) error) error {
+	// Get list of shards to work on.
+	shardIDs := make([]uint64, 0, len(report.shardPaths))
+	for id := range report.shardPaths {
+		pth := path.Join(report.shardPaths[id], "index")
+
+		report.shardIdxs[id] = tsi1.NewIndex(report.sfile,
+			"",
+			tsi1.WithPath(pth),
+			tsi1.DisableCompactions(),
+		)
+
+		// Initialise cardinality set to store cardinalities for each shard
+		report.cardinalities[id] = map[string]*cardinality{}
+
+		shardIDs = append(shardIDs, id)
+	}
+
+	errC := make(chan error, len(shardIDs))
+	var maxi uint32 // index of maximum shard being worked on.
+	for k := 0; k < report.concurrency; k++ {
+		go func() {
+			for {
+				i := int(atomic.AddUint32(&maxi, 1) - 1) // Get next shard to work on.
+				if i >= len(shardIDs) {
+					return // No more work.
+				}
+				errC <- fn(shardIDs[i])
+			}
+		}()
+	}
+
+	// Check for error
+	for i := 0; i < cap(errC); i++ {
+		if err := <-errC; err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Cardinality struct and methods
+type cardinality struct {
+	name  []byte
+	short []uint32
+	set   *tsdb.SeriesIDSet
+}
+
+func (c *cardinality) add(x uint64) {
+	if c.set != nil {
+		c.set.AddNoLock(x)
+		return
+	}
+
+	c.short = append(c.short, uint32(x)) // Series IDs never get beyond 2^32
+
+	// Cheaper to store in bitmap.
+	if len(c.short) > useBitmapN {
+		c.set = tsdb.NewSeriesIDSet()
+		for _, s := range c.short {
+			c.set.AddNoLock(uint64(s))
+		}
+		c.short = nil
+		return
+	}
+}
+
+func (c *cardinality) cardinality() int64 {
+	if c == nil || (c.short == nil && c.set == nil) {
+		return 0
+	}
+
+	if c.short != nil {
+		return int64(len(c.short))
+	}
+	return int64(c.set.Cardinality())
+}
+
+type cardinalities []*cardinality
+
+func (a cardinalities) Len() int           { return len(a) }
+func (a cardinalities) Less(i, j int) bool { return a[i].cardinality() < a[j].cardinality() }
+func (a cardinalities) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func (report *reportTSI) cardinalityByMeasurement(shardID uint64) error {
+	if err := report.shardIdxs[shardID].Open(); err != nil {
+		return err
+	}
+
```

Comment:
I would expect the `defer report.shardIdxs[shardID].Close()` to be here. Is that where you had it when you saw the errors?

Reply:
I think I figured out why the "index is closing" error was happening when the close was here. When `report.printSummaryByMeasurement()` is called on line 121 (after calling `report.calculateCardinalities(report.cardinalityByMeasurement`), it expects the indexes to still be open, otherwise it returns an error on line 346. It cannot return the measurement iterator without the indexes being open at that step. 

I didn't see a clean way to close each index individually when they are done being used, because `MeasurementIterator()` is called on all indexes at the same time. I decided to close them all after `report.printSummaryByMeasurement()` returns (there was already a for loop there that I could add the close to).  Let me know if you see a better place to close them.